### PR TITLE
Add example for using HK2 for DI in your project

### PR DESCRIPTION
**Pull request by kibertoad**
refs #1626

###### Problem:
This provides a sample documentation for how HK2 that comes included with Dropwizard can be used for implementing test-friendly DI mechanism in your application.

###### Solution:
Added new section "Dropwizard Dependency Injection" to the documentation.

### CODE REVIEW

**Comment by kibertoad** [comment]
I'll also create an alternate proposal which uses Application rather than Configuration for parametrizing DI in tests, in case that is preferable.

**Review by joschi — APPROVED** [review]

**Comment by joschi** [comment]
@kibertoad Thanks a lot for your work on this PR!

I'll merge it now and any polishing can be done in separate PRs.

File: docs/source/manual/di.rst [hunk 1]
```diff
@@ -0,0 +1,158 @@
+.. _man-di:
+
+################################
+Dropwizard Dependency Injection
+################################
+
+.. highlight:: text
+
+.. rubric:: Dropwizard provides you with simple dependency injection mechanism, using HK2,
+            out-of-the-box, and you can add support for more advanced DI by using Guice bundle.
```
Comment (OneCricketeer) [inline docs/source/manual/di.rst:10]: Would be nice if "Guice Bundle" had some hyperlink. 

Or also mention Dagger2, if people are using that (referenced in the linked issue)

Comment (joschi) [reply docs/source/manual/di.rst:10]: @OneCricketeer Could you create a pull request for this?

Comment (OneCricketeer) [reply docs/source/manual/di.rst:10]: Sure. One question, are we recommending <https://github.com/xvik/dropwizard-guicey> ? I personally haven't used that one, but our DW 1.x projects are [using the Hubspot one, but the maintainer said it was abandoned](https://github.com/HubSpot/dropwizard-guicier/issues/61#issuecomment-586598066).

Also not really aware of a public bundle that actually uses Dagger w/ DW

**Comment by arjantijms** [comment]
I just came across this, and maybe a new issue is warranted, but I noticed the following:

> a CDI-compliant dependency injection framework.

This is not correct. HK2 is a DI-compliant dependency injection framework. DI is the previous JSR330 (AtInject) specification, and is used by CDI, but is not CDI itself. Guice is a DI-compliant implementation as well (the reference implementation actually).

**Pull request merged**